doc/developer/design/20230801_sink_conn_mgmt.md

```diff
@@ -0,0 +1,92 @@
+# Sink connection management refactor
+
+- Associated:
+    - [sentry: panic: Reference to absent collection](https://github.com/MaterializeInc/materialize/issues/17210)
+    - [sinks: bootstrapping catalog does not wait for sink
+      creation](https://github.com/MaterializeInc/materialize/issues/20019)
+
+## Context
+
+The current implementation of sink connections relies on establishing the
+connection in adapter code. In addition to lacking parity with source connection
+management, this also introduces an issue which is detailed in [sinks:
+      bootstrapping catalog does not wait for sink
+      creation](https://github.com/MaterializeInc/materialize/issues/20019).
+
+## Goals
+
+- Detail and defend the approach for moving sink connection management into
+  storage, while still detecting some errors during sink creation.
+
+## Overview
+
+In general, the approach will be to mirror the structure of source connection
+management for sinks. This means:
+
+- Check for potential errors during purification.
+    - This introduces the possibility for TOCTOU errors, so it will be possible
+      to create a sink that we believe will work only to have the sink
+      immediately fail.
+- Create the sink in the storage controller after purification or during
+  bootstrapping, i.e. remove the notion of
+  `prepare_export`/`SinkConnectionReady`.
+    - This guarantees that we understand when connections to sinks are
+      established and does not potentially leave them in a dangling state during
+      bootstrapping.
```

Comment:
Wait, sorry, so is the topic getting created by the storage controller or by clusterd?

Reply:
In this proposal, it's `clusterd`; this mirrors how we e.g. create replication slots in PG sources.